aws/resource_aws_rds_cluster_instance_test.go

```diff
@@ -872,28 +872,70 @@ func TestAccAWSRDSClusterInstance_CACertificateIdentifier(t *testing.T) {
 	})
 }
 
+func testAccPreCheckAWSRDSClusterInstance_performanceInsights(t *testing.T) {
+	conn := testAccProvider.Meta().(*AWSClient).rdsconn
+
+	input := &rds.DescribeOrderableDBInstanceOptionsInput{
+		Engine:        aws.String("mysql"),
+		EngineVersion: aws.String("8.0.20"), // version cuts response time
+	}
```

Comment:
This is a nice little setup. I think the challenge here is that the configurations themselves are using different RDS engines and versions, which this check is skipping testing when it shouldn't. For example:

```
=== CONT  TestAccAWSRDSClusterInstance_PerformanceInsightsKmsKeyId_AuroraMysql1
    resource_aws_rds_cluster_instance_test.go:903: PreCheck fail: performance insights not supported, skipping acceptance test
--- SKIP: TestAccAWSRDSClusterInstance_PerformanceInsightsKmsKeyId_AuroraMysql1 (2.43s)
```

Should pass. Checking via CLI using the proper `aurora` engine:

```console
$ aws rds describe-orderable-db-instance-options --engine aurora
        {
            "Engine": "aurora",
            "EngineVersion": "5.6.mysql_aurora.1.23.0",
            "DBInstanceClass": "db.r5.large",
            "LicenseModel": "general-public-license",
            "AvailabilityZones": [
                {
                    "Name": "us-west-2a"
                },
                {
                    "Name": "us-west-2b"
                },
                {
                    "Name": "us-west-2c"
                },
                {
                    "Name": "us-west-2d"
                }
            ],
            "MultiAZCapable": false,
            "ReadReplicaCapable": false,
            "Vpc": true,
            "SupportsStorageEncryption": true,
            "StorageType": "aurora",
            "SupportsIops": false,
            "SupportsEnhancedMonitoring": true,
            "SupportsIAMDatabaseAuthentication": true,
            "SupportsPerformanceInsights": true,
            "AvailableProcessorFeatures": [],
            "SupportedEngineModes": [
                "provisioned"
            ],
            "SupportsKerberosAuthentication": false,
            "OutpostCapable": false,
            "SupportsGlobalDatabases": true
        },
```

For starters, I would suggest forcing these API parameters to be passed in as function parameters:

```suggestion
func testAccPreCheckAWSRDSClusterInstance_performanceInsights(t *testing.T, engine string, engineVersion string) {
	conn := testAccProvider.Meta().(*AWSClient).rdsconn

	input := &rds.DescribeOrderableDBInstanceOptionsInput{
		Engine:        aws.String(engine),
		EngineVersion: aws.String(engineVersion), // version cuts response time
	}
```

With each of the tests implementing the expected parameters, for example:

```
// TestAccAWSRDSClusterInstance_PerformanceInsightsKmsKeyId_AuroraMysql1
testAccPreCheckAWSRDSClusterInstance_performanceInsights(t, "aurora", "5.6.mysql_aurora.1.23.0")

// TestAccAWSRDSClusterInstance_PerformanceInsightsKmsKeyId_AuroraMysql2
testAccPreCheckAWSRDSClusterInstance_performanceInsights(t, "aurora-mysql", "5.7.mysql_aurora.2.08.2")

// TestAccAWSRDSClusterInstance_PerformanceInsightsKmsKeyId_AuroraPostgresql
testAccPreCheckAWSRDSClusterInstance_performanceInsights(t, "aurora-postgresql", "11.8")
```

When some of those do begin running the tests, the data source may need to be added/adjusted to pass.

If this seems like too much for this PR, we can always do it separately. 👍 

---

This type of PreCheck logic may represent a good case for separately enhancing the acceptance testing framework `TestStep` to accept an error regular expression to skip a test, so external logic like this is not necessarily required. There's more to noodle on this idea though -- while it would prevent this external-to-the-configuration value copy-pasting and API lookup, but it would potentially trigger false positive skipping of tests if the API or configuration changes in unexpected ways. 🤔  For example, if a newer version of an engine was made available but it didn't immediately support performance insights, then the test would skip rather than fail as we would currently expect, so maybe hardcoding/choosing some preferred values would still be the way to go.

Reply:
I really like this idea. I'll create a tracking issue for it. In your example, if the new engine didn't support immediately, but did eventually, the precheck would just initially skip but then eventually not skip. I think it would work well in enough situations that it would be a big improvement.

Reply:
@bflad See if #15066 captures your idea.

Reply:
FYI On GovCloud, performance insights are not supported for any engine or any version.

Reply:
No engine/version combo on GovCloud supports performance insights currently:

```
% aws rds describe-orderable-db-instance-options --engine aurora --query "OrderableDBInstanceOptions[] [SupportsPerformanceInsights]" --output text | sort -u
aws rds describe-orderable-db-instance-options --engine aurora-mysql --query "OrderableDBInstanceOptions[] [SupportsPerformanceInsights]" --output text | sort -u
aws rds describe-orderable-db-instance-options --engine aurora-postgresql --query "OrderableDBInstanceOptions[] [SupportsPerformanceInsights]" --output text | sort -u
aws rds describe-orderable-db-instance-options --engine docdb --query "OrderableDBInstanceOptions[] [SupportsPerformanceInsights]" --output text | sort -u
aws rds describe-orderable-db-instance-options --engine mariadb --query "OrderableDBInstanceOptions[] [SupportsPerformanceInsights]" --output text | sort -u
aws rds describe-orderable-db-instance-options --engine mysql --query "OrderableDBInstanceOptions[] [SupportsPerformanceInsights]" --output text | sort -u
aws rds describe-orderable-db-instance-options --engine neptune --query "OrderableDBInstanceOptions[] [SupportsPerformanceInsights]" --output text | sort -u
aws rds describe-orderable-db-instance-options --engine oracle-ee --query "OrderableDBInstanceOptions[] [SupportsPerformanceInsights]" --output text | sort -u
aws rds describe-orderable-db-instance-options --engine oracle-se --query "OrderableDBInstanceOptions[] [SupportsPerformanceInsights]" --output text | sort -u
aws rds describe-orderable-db-instance-options --engine oracle-se1 --query "OrderableDBInstanceOptions[] [SupportsPerformanceInsights]" --output text | sort -u
aws rds describe-orderable-db-instance-options --engine oracle-se2 --query "OrderableDBInstanceOptions[] [SupportsPerformanceInsights]" --output text | sort -u
aws rds describe-orderable-db-instance-options --engine postgres --query "OrderableDBInstanceOptions[] [SupportsPerformanceInsights]" --output text | sort -u
aws rds describe-orderable-db-instance-options --engine sqlserver-ee --query "OrderableDBInstanceOptions[] [SupportsPerformanceInsights]" --output text | sort -u
aws rds describe-orderable-db-instance-options --engine sqlserver-ex --query "OrderableDBInstanceOptions[] [SupportsPerformanceInsights]" --output text | sort -u
aws rds describe-orderable-db-instance-options --engine sqlserver-se --query "OrderableDBInstanceOptions[] [SupportsPerformanceInsights]" --output text | sort -u
aws rds describe-orderable-db-instance-options --engine sqlserver-web --query "OrderableDBInstanceOptions[] [SupportsPerformanceInsights]" --output text | sort -u

False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
False
%
```

Reply:
There are now 2 prechecks. One takes just the engine and then finds the default version and the other takes both the engine and the version. Of the 9 performance insights tests, 6 use the default version and 3 had a hardcoded version in the config. I'm now passing the engine and version, where applicable, to the config so that they don't get changed in one place but not the other.